### PR TITLE
mkldnn activations.relu

### PR DESCRIPTION
**Pull request by sreekanth-yalachigere**
Activator Relu is implemented using MKLDNN.

```
core/providers/mkldnn/activation
    activations.cc
    activations.h
```

Modified CPU Relu class definition to enable MKLDNN Relu class to inherit for fallback option

`class Relu final : public OpKernel {  `
to 
`class Relu : public OpKernel { `

### CODE REVIEW

File: onnxruntime/core/providers/mkldnn/activation/activations.cc [hunk 1]
```diff
@@ -0,0 +1,238 @@
+/* Copyright(C) 2018 Intel Corporation
```
Comment (pranavsharma) [inline onnxruntime/core/providers/mkldnn/activation/activations.cc]: Is there a legal reason to use this license?

File: onnxruntime/core/providers/mkldnn/activation/activations.cc [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+
+  mkldnn::memory::dims src_dims_mkl(x_dims.begin(), x_dims.end());
+  mkldnn::memory::dims dst_dims_mkl(y_dims.begin(), y_dims.end());
+
+  try {
+    ReluParams pool_params(src_dims_mkl, dst_dims_mkl, x_shape.NumDimensions());
+    ReluPrimitive<T>* relulPrimitive = ReluPrimitivePool<T>::Get(pool_params);
+
+    relulPrimitive->Compute(src_data, dst_data);
+  } catch (mkldnn::error& e) {
```
Comment (pranavsharma) [inline onnxruntime/core/providers/mkldnn/activation/activations.cc]: const mkldnn::error&

File: onnxruntime/core/providers/mkldnn/activation/activations.h [hunk 3]
```diff
@@ -0,0 +1,32 @@
+/* Copyright(C) 2018 Intel Corporation
```
Comment (pranavsharma) [inline onnxruntime/core/providers/mkldnn/activation/activations.h]: same comment as above for the license

File: onnxruntime/core/providers/mkldnn/activation/activations.cc (hunk 1 above)
Comment (pranavsharma) [reply onnxruntime/core/providers/mkldnn/activation/activations.cc]: I believe this should be ok. Will double check.

File: onnxruntime/core/providers/mkldnn/activation/activations.cc [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+  ReluParams(mkldnn::memory::dims& src_dims, mkldnn::memory::dims& dst_dims, 
+		size_t dimensions = 0)
+      : src_dims(src_dims),
+        dst_dims(dst_dims),
+        num_dimensions(dimensions) {}
+
+  // Used as the key for Pool Primitive Reuse Pool.
+  std::string ToString() const {
+    std::string key;
+    key.reserve(128);
```
Comment (jywu-msft) [inline onnxruntime/core/providers/mkldnn/activation/activations.cc]: probably don't need 128 bytes for just a src and dst dims.

File: onnxruntime/core/providers/mkldnn/activation/activations.cc [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+	  context_.dst_mem->set_data_handle(
+		  static_cast<void*>(const_cast<T*>(dst_data)));
+	  context_.stream->submit(context_.net);
+
+	  context_.src_mem->set_data_handle(nullptr);
+	  context_.dst_mem->set_data_handle(nullptr);
+	  return;
+  }
+
+  mkldnn::memory::format GetSrcMemoryFormat() const { return context_.src_fmt; }
```
Comment (jywu-msft) [inline onnxruntime/core/providers/mkldnn/activation/activations.cc]: GetSrcMemoryFormat and GetDstMemoryFormat are not used?
can get rid of the members (src_fmt, dst_fmt) in context as well then.

File: onnxruntime/core/providers/mkldnn/activation/activations.cc [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+	GetDstMemoryDesc() const { return context_.dst_md; }
+
+  std::unique_ptr<mkldnn::eltwise_forward::primitive_desc> 
+	GetPrimitiveDesc() const {
+		return context_.relu_fwd_pd;
+  }
+
+ private:
+  struct ReluContext {
+    mkldnn::memory::format src_fmt;
```
Comment (jywu-msft) [inline onnxruntime/core/providers/mkldnn/activation/activations.cc]: per comment above. remove src_fmt, and dst_fmt if no need for them.

File: onnxruntime/core/providers/mkldnn/activation/activations.cc [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+      new mkldnn::eltwise_forward(*context_.relu_fwd_pd, *context_.src_mem, 
+			*context_.dst_mem));
+    context_.net.push_back(*context_.relu_fwd);
+  }
+
+  ReluContext context_;
+  mkldnn::engine& cpu_engine_;
+};
+
+// Pool which allows for reuse of MKLDNN Conv2d primitives which are expensive
```
[the remark this answers is not in the document]
Comment (sreekanth-yalachigere) [reply onnxruntime/core/providers/mkldnn/activation/activations.cc]: Changing `Conv2d `to `Relu`

File: onnxruntime/core/providers/mkldnn/activation/activations.cc (hunk 4 above)
Comment (sreekanth-yalachigere) [reply onnxruntime/core/providers/mkldnn/activation/activations.cc]: will change this to 64 bytes

File: onnxruntime/core/providers/mkldnn/activation/activations.cc (hunk 5 above)
Comment (sreekanth-yalachigere) [reply onnxruntime/core/providers/mkldnn/activation/activations.cc]: Removed

File: onnxruntime/core/providers/mkldnn/activation/activations.cc (hunk 6 above)
Comment (sreekanth-yalachigere) [reply onnxruntime/core/providers/mkldnn/activation/activations.cc]: removed

**Comment by sreekanth-yalachigere** [comment]
@pranavsharma committed the following. 
Modified the license header

added _const_ at catch

`catch (const mkldnn::error& e)`

**Comment by pranavsharma** [comment]
I still don't see MIT license mentioned like other files. You should mention ```// Licensed under the MIT License.``` at the top of the file.

**Comment by sreekanth-yalachigere** [comment]
@pranavsharma added second line in header

`// Licensed under the MIT License.
`

**Comment by pranavsharma** [comment]
We did some build pipeline changes recently. Can you please rebase your changes? Thanks.

**Comment by sreekanth-yalachigere** [comment]
@pranavsharma rebase done.

File: onnxruntime/core/providers/mkldnn/activation/activations.cc [hunk 8]
```diff
@@ -0,0 +1,223 @@
...
+    context_.net.push_back(*context_.relu_fwd);
+  }
+
+  ReluContext context_;
+  mkldnn::engine& cpu_engine_;
+};
+
+// Pool which allows for reuse of MKLDNN Relu primitives which are expensive 
+// to instantiate. To address thread safety, the primitives are stored in a map 
+// on thread local storage.
```
Comment (snnn) [inline onnxruntime/core/providers/mkldnn/activation/activations.cc]: Who is in the thread local storage?

Comment (sreekanth-yalachigere) [reply onnxruntime/core/providers/mkldnn/activation/activations.cc]: @snnn kernel objects like Relu lives in thread local storage. This is very similar to what @weixingzhang's implementation of Conv and Pool kernels.

**Pull request merged**